### PR TITLE
executor: add unit tests for prepared statements.

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

File: executor/exeutor_test.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor_test
```
Comment (qiuyesuifeng) [inline executor/exeutor_test.go]: Rename filename to `executor_test. go`.

File: executor/exeutor_test.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	tk.MustExec("insert prepare_test (c1) values (1),(2),(NULL)")
+
+	tk.MustExec(`prepare stmt_test_1 from 'select id from prepare_test where id > ?'; set @a = 1; execute stmt_test_1 using @a;`)
+	tk.MustExec(`prepare stmt_test_2 from 'select 1'`)
+	// Prepare multiple statement is not allowed.
+	_, err := tk.Exec(`prepare stmt_test_3 from 'select id from prepare_test where id > ?;select id from prepare_test where id > ?;'`)
+	c.Assert(executor.ErrPrepareMulti.Equal(err), IsTrue)
+	// The variable count do not match.
+	_, err = tk.Exec(`prepare stmt_test_4 from 'select id from prepare_test where id > ? and id < ?'; set @a = 1; execute stmt_test_4 using @a;`)
+	c.Assert(executor.ErrWrongParamCount.Equal(err), IsTrue)
```
Comment (qiuyesuifeng) [inline executor/exeutor_test.go]: How about ErrInvalidParamCount?

File: executor/exeutor_test.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	tk.MustExec("drop table if exists prepare_test")
+	tk.MustExec("create table prepare_test (id int PRIMARY KEY AUTO_INCREMENT, c1 int, c2 int, c3 int default 1)")
+	tk.MustExec("insert prepare_test (c1) values (1),(2),(NULL)")
+
+	tk.MustExec(`prepare stmt_test_1 from 'select id from prepare_test where id > ?'; set @a = 1; execute stmt_test_1 using @a;`)
+	tk.MustExec(`prepare stmt_test_2 from 'select 1'`)
+	// Prepare multiple statement is not allowed.
+	_, err := tk.Exec(`prepare stmt_test_3 from 'select id from prepare_test where id > ?;select id from prepare_test where id > ?;'`)
+	c.Assert(executor.ErrPrepareMulti.Equal(err), IsTrue)
+	// The variable count do not match.
```
Comment (qiuyesuifeng) [inline executor/exeutor_test.go]: Does not match.

File: executor/exeutor_test.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	_, err = tk.Exec(`prepare stmt_test_4 from 'select id from prepare_test where id > ? and id < ?'; set @a = 1; execute stmt_test_4 using @a;`)
+	c.Assert(executor.ErrWrongParamCount.Equal(err), IsTrue)
+	// Prepare and deallocate prepared statement immediately.
+	tk.MustExec(`prepare stmt_test_5 from 'select id from prepare_test where id > ?'; deallocate prepare stmt_test_5;`)
+
+	// Not found
```
Comment (qiuyesuifeng) [inline executor/exeutor_test.go]: Not exist prepare stmt.

File: executor/exeutor_test.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+	// Use parameter marker with argument will run prepared statement.
+	result := tk.MustQuery("select distinct c1, c2 from prepare_test where c1 = ?", 1)
+	result.Check([][]interface{}{{1, nil}})
+
+	// Call Session PrepareStmt directly to get stmtId.
+	stmtId, _, _, err := tk.Se.PrepareStmt("select c1, c2 from prepare_test where c1 = ?")
+	c.Assert(err, IsNil)
+	_, err = tk.Se.ExecutePreparedStmt(stmtId, 1)
+	c.Assert(err, IsNil)
```
Comment (qiuyesuifeng) [inline executor/exeutor_test.go]: Please check result.

File: executor/exeutor_test.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+	// Drop a column so the prepared statement become invalid.
+	tk.MustExec("alter table prepare_test drop column c2")
+
+	// There should be schema changed error.
+	_, err = tk.Se.ExecutePreparedStmt(stmtId, 1)
+	c.Assert(executor.ErrSchemaChanged.Equal(err), IsTrue)
+
+	// Coverage.
+	exec := &executor.ExecuteExec{}
```
Comment (qiuyesuifeng) [inline executor/exeutor_test.go]: A little strange for ExecuteExec name, seems like exec exec...

Comment (coocood) [inline executor/exeutor_test.go]: It follows the naming pattern for executors.

File: executor/exeutor_test.go (hunk 2 above)
Comment (coocood) [inline executor/exeutor_test.go]: I thing invalid is all about the count itself, but wrong means it is different than the count of the prepared statement.

File: executor/exeutor_test.go (hunk 5 above)
Comment (coocood) [inline executor/exeutor_test.go]: The same statement has been checked before, but with a higher level testKit, the purpose of this code is to test the schema change error, we need to get the statementId, checking the result at low level is tedious.

**Comment by shenli** [comment]
LGTM

**Pull request merged**